staker-rewards/lib/sla-calculator.js

```diff
@@ -0,0 +1,134 @@
+
+export default class SLACalculator {
+    constructor(
+        openedKeeps,
+        keygenFailKeeps,
+        deactivatedKeeps,
+        signatureFailKeeps
+    ) {
+        this.openedKeeps = openedKeeps
+        this.keygenFailKeeps = keygenFailKeeps
+        this.deactivatedKeeps = deactivatedKeeps
+        this.signatureFailKeeps = signatureFailKeeps
+    }
+
+    static initialize(cache, interval) {
+        const isInInterval = (timestamp) =>
+            interval.start <= timestamp && timestamp < interval.end
+
+        // Step 1 of keygen SLA: get keeps opened within the given interval
+        const openedKeeps = cache
+            .getKeeps() // get keeps with all statuses
+            .filter(keep =>
+                // get keeps whose creation timestamps are within
+                // the given interval
+                isInInterval(keep.creationTimestamp)
+            )
+
+        // Step 2 of keygen SLA: from keeps opened within the given interval,
+        // get the ones which have been eventually terminated due to keygen fail
+        const keygenFailKeeps = openedKeeps
+            .filter(keep =>
+                keep.status.name === KeepStatus.TERMINATED &&
+                keep.status.cause === KeepTerminationCause.KEYGEN_FAIL
+            )
+
+        // Step 1 of signature SLA: get keeps closed within the given interval
+        const closedKeeps = cache
+            .getKeeps(KeepStatus.CLOSED) // get closed keeps
+            .filter(keep =>
+                // get keeps whose statuses have been changed within the
+                // given interval
+                isInInterval(keep.status.timestamp)
+            )
+
+        // Step 2 of signature SLA: get keeps terminated due to signature fail,
+        // within the given interval
+        const signatureFailKeeps = cache
+            .getKeeps(KeepStatus.TERMINATED) // get terminated keeps
+            .filter(keep =>
+                // get keeps whose statuses have been changed within the
+                // given interval
+                isInInterval(keep.status.timestamp)
+            )
+            .filter(keep =>
+                // get keeps which have been terminated due to signature fail
+                keep.status.cause === KeepTerminationCause.SIGNATURE_FAIL
+            )
+
+        // Step 3 of signature SLA: Concatenate keeps closed and terminated
+        // due to signature fail, within the given interval. This way
+        // we obtain an array of keeps whose statuses have been changed from
+        // `active` to `closed` or from `active` to `terminated` due to
+        // signature fails. Implicitly, this means a keep became not active
+        // due to one of the following causes:
+        // - keep has been closed after delivering a signature successfully
+        // - keep has been terminated after not delivering a signature
+        const deactivatedKeeps = [].concat(closedKeeps, signatureFailKeeps)
```

Comment:
There is one edge case we missed. Keep can be closed by the application even though it was never asked to provide a signature. In the case of tBTC, this is a funding timeout scenario. For example, if we generate mainnet report from the moment of deploying contracts and look at the operator `0xF1De9490Bf7298b5F350cE74332Ad7cf8d5cB181`, we see this operator has `signatureCount = 159`:

![image](https://user-images.githubusercontent.com/4712360/100736598-4a253300-33d3-11eb-89b5-06d1edb91e0f.png)


It does not seem right as a significant number of `closedKeeps` [are funding time-out keeps](https://allthekeeps.com/operator/0xf1de9490bf7298b5f350ce74332ad7cf8d5cb181) that were never asked for a signature. 

I think the fix should be to add a function to `keep.js` that will let us say if the given keep was ever asked to provide a signature (one occurrence of `SignatureRequested` is enough) and if not, filter out such keeps. For example:

```
const deactivatedAndAskedForSigning = deactivatedKeeps.filter(keep => wasAskedForSignature(keep))
```

Reply:
Ok, done. Unfortunately, this change increases the script execution time significantly. For now, I don't see an easy way to optimize those checks.

Reply:
Yeah, agreed. We have to live with it for now. In the future, we could be caching if a keep was asked for a signature refreshing all cases when the keep wasn't asked. Probably, adding some optimization with interval start/end date and whether there was a chance the signature was requested (e.g. it is not a case for closed keeps before interval start date).

Let's leave this discussion open for now - maybe we can implement some optimization in a separate PR at the very end, once we'll have all other pieces working together.